app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java

```diff
@@ -66,7 +68,7 @@
 
     private Lazy<ApiClient> apiClient = inject(ApiClient.class);
     private Lazy<PlaybackManager> playbackManager = inject(PlaybackManager.class);
-    private Lazy<UserPreferences> userPreferences = inject(UserPreferences.class);
+    private Lazy<PreferenceStore> userPreferences = inject(PreferenceStore.class);
```

Comment:
We use the UserPreferences store so use that one, not an interface.

Reply:
But this circles back to breaking mocks:
By specifying a concrete type, our mock also has to pull in the parent class when generics are involved. The abstract base class then tries to interact with a real preferences store (which is null as we're in a JVM unit test) and fails.

Injecting an interface allows us to still use all the same members, but removes the need for the mocking lib to inherit from the parent class

Reply:
But that's a problem to solve in the tests, not in the preferences library.

Reply:
I'm going to push for this. 

I understand I'm simply a contributor and I'm not trying to do anything hostile / force my approach onto this project in any way, hence I'm happy to make as many changes as required or wanted. But in this case using interfaces instead of implementations is a widely accepted pattern used in software development to make things more testable:

General:
https://stackoverflow.com/questions/2697783/what-does-program-to-interfaces-not-implementations-mean
https://stackoverflow.com/questions/3194278/should-you-always-code-to-interfaces-in-java
https://betterprogramming.pub/code-against-interfaces-not-implementations-37b30e7ab992

More relevant to this case with relation to mocks:
https://softwareengineering.stackexchange.com/questions/159813/do-i-need-to-use-an-interface-when-only-one-class-will-ever-implement-it

And the main server repo uses this pattern to generate / inject mocks too (I simply searched `interface` and grabbed the latest PR at random) :
https://github.com/jellyfin/jellyfin/pull/7039/files

----

Long-term I'd like to help write more tests for this project to help with some of the stability problems people encounter on the diverse (and sometimes painful) Android TV ecosystem.

Whilst Android development isn't my speciality, my hope is adding better / new tests helps maintainers at least know that things they fix don't regress, giving you more time to focus on the things I'm not qualified to help with i.e. Playback rewrite.

But this either requires significant architectural changes with a factory / injection focus for testability or small shims and simple interfaces on existing code, which I would argue is better than nothing.

Reply:
I get your point, I too believe we could use more tests in the app. Unfortunately the app is still 80% legacy code (pretty much all Java is) and a lot of the newly written Kotlin code must be kept compatible with the old code. That's why some weird architectural choices were made. This is also why I personally do not create tests right now.

I'm not against interfaces. But in this particular situation it just does not work. The preferences library is not made to work like this and would need more significant changes to be properly testable. The type that a preference is read from **must** be the same as where the preference definition comes from, this is for various safety reasons. The Kotlin compiler doesn't check for this right now due to the way I implemented the system. If I'd rewritten the preferences code today I would use a different approach and architecture for it.

For now, I'd propose dropping the tests from this PR and moving them to a new one. That way the speed controls can be merged as-is and we can further discuss changes needed for these specific tests. It would also make it easier to review, just like when we moved the addition of Mockk etc. to separate PR's.

Reply:
> The Kotlin compiler doesn't check for this right now due to the way I implemented the system.

Sorry, I completely misread your previous comment as we won't take this change, rather than this is a technical limitation around how types are implemented.

> ...I'd propose dropping the tests from this PR and moving them to a new one...
I completely agree about getting this in rather than stalling out. I would argue we can simply drop the PlaybackController tests all-together until the rewrite is done. Since trying to figure out how to make everything play nice, then simply ripping it out later on seems like a waste.

Also I have to say, thank you for your patience with me so far whilst I try to get the hang of this